clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -0,0 +1,1021 @@
+     *
+     * @param throwable                         The exception.
+     *
+     * @return                                  A compact human-readable string.
+     */
+    static String prettyPrintException(Throwable throwable) {
+        if (throwable == null)
+            return "Null exception.";
+        if (throwable.getMessage() != null) {
+            return throwable.getClass().getSimpleName() + ": " + throwable.getMessage();
+        }
+        return throwable.getClass().getSimpleName();
+    }
+
+    /**
+     * Close an AutoCloseable, suppressing any exceptions.
+     *
+     * @param what                              Human-readable string explaining what we are closing.
+     * @param closeable                         What we are closing, or null to ignore.
+     */
+    static void closeQuietly(String what, AutoCloseable closeable) {
+        if (closeable == null)
+            return;
+        try {
+            closeable.close();
+        } catch (Throwable exc) {
+            log.error("Error closing " + what, exc);
+        }
+    }
+
+    static KafkaAdminClient create(AdminClientConfig config) {
+        Metadata metadata = null;
+        Metrics metrics = null;
+        NetworkClient networkClient = null;
+        Time time = Time.SYSTEM;
+        String clientId = generateClientId(config);
+        ChannelBuilder channelBuilder = null;
+        Selector selector = null;
+        ApiVersions apiVersions = new ApiVersions();
+
+        try {
+            metadata = new Metadata(config.getLong(AdminClientConfig.RETRY_BACKOFF_MS_CONFIG),
+                    config.getLong(AdminClientConfig.METADATA_MAX_AGE_CONFIG));
+            metrics = new Metrics(new MetricConfig(), new LinkedList<MetricsReporter>(), time);
+            String metricGrpPrefix = "adminclient";
+            channelBuilder = ClientUtils.createChannelBuilder(config);
+            selector = new Selector(config.getLong(AdminClientConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG),
+                    metrics, time, metricGrpPrefix, channelBuilder);
+            networkClient = new NetworkClient(
+                selector,
+                metadata,
+                clientId,
+                100,
+                config.getLong(AdminClientConfig.RECONNECT_BACKOFF_MS_CONFIG),
+                config.getInt(AdminClientConfig.SEND_BUFFER_CONFIG),
+                config.getInt(AdminClientConfig.RECEIVE_BUFFER_CONFIG),
+                config.getInt(AdminClientConfig.REQUEST_TIMEOUT_MS_CONFIG),
+                time,
+                true,
+                apiVersions);
+            channelBuilder = null;
+            return new KafkaAdminClient(config, clientId, time, metadata, metrics, networkClient);
+        } catch (Throwable exc) {
+            closeQuietly("Metrics", metrics);
+            closeQuietly("NetworkClient", networkClient);
+            closeQuietly("Selector", selector);
+            closeQuietly("ChannelBuilder", channelBuilder);
```

Comment:
The constructor is private and can't be called from outside this class.  Even the create methods are package-private and only used by the public interface in AdminClient and by unit tests.  Putting a lot of logic in the constructor doesn't work very well because the logic needs to be different for the unit tests (which don't use NetworkClient, Selector, or ChannelBuilder, for example)